middle_end/flambda/types/template/flambda_type.templ.ml

```diff
@@ -741,7 +741,120 @@ let prove_strings env t : String_info.Set.t proof =
   | Naked_immediate _ | Naked_float _ | Naked_int32 _ | Naked_int64 _
   | Naked_nativeint _ -> wrong_kind ()
 
-let prove_block_field_simple env ~min_name_mode t field_index : Simple.t proof =
+let prove_is_tagging_of_simple
+    ~prove_function env ~min_name_mode t : Simple.t proof =
+  let wrong_kind () =
+    Misc.fatal_errorf "Kind error: expected [Value]:@ %a" print t
+  in
+  match expand_head t env with
+  | Const (Tagged_immediate imm) ->
+    Proved (Simple.const (Reg_width_const.naked_immediate imm))
+  | Value (Ok (Variant { immediates; blocks; is_unique = _; })) ->
+    begin match blocks with
+    | Unknown -> Unknown
+    | Known blocks ->
+      match prove_function with
+      | `Prove_is_always_tagging_of_simple
+        when not (Row_like.For_blocks.is_bottom blocks) ->
+        Unknown
+      | `Prove_is_always_tagging_of_simple
+        (* when (Row_like.For_blocks.is_bottom blocks) *)
+      | `Prove_could_be_tagging_of_simple ->
+        match immediates with
+        | Unknown -> Unknown
+        | Known t ->
+          let from_alias =
+            match get_alias_exn t with
+            | simple ->
+              begin match
+                Typing_env.get_canonical_simple_exn env ~min_name_mode simple
+              with
+              | simple -> Some simple
+              | exception Not_found -> None
+              end
+            | exception Not_found -> None
+          in
+          match from_alias with
+          | Some simple -> Proved simple
+          | None ->
+            match prove_naked_immediates env t with
+            | Unknown -> Unknown
+            | Invalid -> Invalid
+            | Proved imms ->
+              match Target_imm.Set.get_singleton imms with
+              | Some imm ->
+                Proved (Simple.const (Reg_width_const.naked_immediate imm))
+              | None -> Unknown
+    end
+  | Value Unknown -> Unknown
+  | Value _ -> Invalid
+  | Const _ | Naked_immediate _ | Naked_float _ | Naked_int32 _
+  | Naked_int64 _ | Naked_nativeint _ -> wrong_kind ()
+
+let prove_is_always_tagging_of_simple =
+  prove_is_tagging_of_simple
+    ~prove_function:`Prove_is_always_tagging_of_simple
+
+let prove_could_be_tagging_of_simple =
+  prove_is_tagging_of_simple
+    ~prove_function:`Prove_could_be_tagging_of_simple
+
+let [@inline always] prove_boxed_number_containing_simple
+      ~contents_of_boxed_number env ~min_name_mode t : Simple.t proof =
+  match expand_head t env with
+  | Value (Ok ty_value) ->
+    begin match contents_of_boxed_number ty_value with
+    | None -> Invalid
+    | Some ty ->
+      match get_alias_exn ty with
```

Comment:
```OCaml
match Typing_env.get_canonical_simple_exn env ~min_name_mode (get_alias_exn ty) with
| exceptin Not_found -> Unknown
| simple -> Proved simple
```
Would be equivalent. Not sure it if is better

Reply:
That would probably bring things dangerously close to the dreaded 80 columns limits...

Reply:
done in the last two commits